src/machine/machine_rp2040_pwm.go

```diff
@@ -94,29 +107,29 @@ func (pwm *pwmGroup) peripheral() uint8 {
 	return uint8((uintptr(unsafe.Pointer(pwm)) - uintptr(unsafe.Pointer(rp.PWM))) / 0x14)
 }
 
-// SetPeriod updates the period of this PWM peripheral.
+// SetPeriod updates the period of this PWM peripheral in nanoseconds.
 // To set a particular frequency, use the following formula:
 //
 //     period = 1e9 / frequency
 //
-// If you use a period of 0, a period that works well for LEDs will be picked.
+// Where frequency is in hertz. If you use a period of 0, a period
+// that works well for LEDs will be picked.
 //
-// SetPeriod will not change the prescaler, but also won't change the current
-// value in any of the channels. This means that you may need to update the
-// value for the particular channel.
-//
-// Note that you cannot pick any arbitrary period after the PWM peripheral has
-// been configured. If you want to switch between frequencies, pick the lowest
-// frequency (longest period) once when calling Configure and adjust the
-// frequency here as needed.
+// SetPeriod will try not to modify TOP if possible to reach the target period.
+// If the period is unattainable with current TOP SetPeriod will modify TOP
+// by the bare minimum to reach the target period. For periods above 130ms
+// phase correct should be enabled by the user before calling SetPeriod, as SetPeriod
+// takes it into account but does not enable it to reach larger periods.
 func (p *pwmGroup) SetPeriod(period uint64) error {
-	if period > 0xffff_ffff {
-		return ErrPeriodTooBig
-	}
 	if period == 0 {
 		period = 1e5
 	}
-	p.setPeriod(period)
+	// Maximum Period is 268369920ns on rp2040, given by (16*255+15)*8*(1+0xffff)*(1+1)/16
+	// With no phase shift max period is half of this value.
+	if period > 268e9 || period < 8 {
```

Comment:
I think it would be easier to move this check into `setPeriod`. That code already contains the code to calculate the maximum possible period and doing it here can introduce bugs. For example, it doesn't look like this code accounts for phase correct PWM and thus is not complete.